crates/wasmtime/src/runtime/gc/enabled/structref.rs

```diff
@@ -0,0 +1,757 @@
+            .with_context(|| format!("cannot set field {index}: type mismatch"))?;
+
+        structref.write_field(&mut store, &layout, field_ty.element_type(), index, value)
+    }
+
+    pub(crate) fn type_index(&self, store: &StoreOpaque) -> Result<VMSharedTypeIndex> {
+        let gc_ref = self.inner.unchecked_try_gc_ref(store)?;
+        let header = store.gc_store()?.header(gc_ref);
+        debug_assert!(header.kind().matches(VMGcKind::StructRef));
+        Ok(header.ty().expect("structrefs should have concrete types"))
+    }
+
+    /// Create a new `Rooted<StructRef>` from the given GC reference.
+    ///
+    /// `gc_ref` should point to a valid `structref` and should belong to the
+    /// store's GC heap. Failure to uphold these invariants is memory safe but
+    /// will lead to general incorrectness such as panics or wrong results.
+    pub(crate) fn from_cloned_gc_ref(
+        store: &mut AutoAssertNoGc<'_>,
+        gc_ref: VMGcRef,
+    ) -> Rooted<Self> {
+        debug_assert!(!gc_ref.is_i31());
+        Rooted::new(store, gc_ref)
+    }
+}
+
+unsafe impl WasmTy for Rooted<StructRef> {
+    #[inline]
+    fn valtype() -> ValType {
+        ValType::Ref(RefType::new(false, HeapType::Struct))
+    }
+
+    #[inline]
+    fn compatible_with_store(&self, store: &StoreOpaque) -> bool {
+        self.comes_from_same_store(store)
+    }
+
+    #[inline]
+    fn dynamic_concrete_type_check(
+        &self,
+        store: &StoreOpaque,
+        _nullable: bool,
+        ty: &HeapType,
+    ) -> Result<()> {
+        match ty {
+            HeapType::Any | HeapType::Eq | HeapType::Struct => Ok(()),
+            HeapType::ConcreteStruct(ty) => self.ensure_matches_ty(store, ty),
+
+            HeapType::Extern
+            | HeapType::NoExtern
+            | HeapType::Func
+            | HeapType::ConcreteFunc(_)
+            | HeapType::NoFunc
+            | HeapType::I31
+            | HeapType::Array
+            | HeapType::ConcreteArray(_)
+            | HeapType::None => bail!(
+                "type mismatch: expected `(ref {ty})`, got `(ref {})`",
+                self._ty(store)?,
+            ),
+        }
+    }
+
+    fn store(self, store: &mut AutoAssertNoGc<'_>, ptr: &mut MaybeUninit<ValRaw>) -> Result<()> {
+        let gc_ref = self.inner.try_clone_gc_ref(store)?;
+        let r64 = gc_ref.as_r64();
+        store.gc_store_mut()?.expose_gc_ref_to_wasm(gc_ref);
+        debug_assert_ne!(r64, 0);
+        let anyref = u32::try_from(r64).unwrap();
+        ptr.write(ValRaw::anyref(anyref));
+        Ok(())
+    }
+
+    unsafe fn load(store: &mut AutoAssertNoGc<'_>, ptr: &ValRaw) -> Self {
+        let raw = ptr.get_anyref();
+        debug_assert_ne!(raw, 0);
+        let gc_ref = VMGcRef::from_r64(raw.into())
+            .expect("valid r64")
+            .expect("non-null");
+        let gc_ref = store.unwrap_gc_store_mut().clone_gc_ref(&gc_ref);
+        StructRef::from_cloned_gc_ref(store, gc_ref)
+    }
```

Comment:
I suspect that this looks the same for anyref/arrayref/etc, do you think this could be refactored into a shared method on `Rooted` perhaps or something like that?

Reply:
(same for the impls below, and of course ok to defer to a future PR if you'd like)

Reply:
Spun out into https://github.com/bytecodealliance/wasmtime/issues/8942 but yeah you're right, this is all basically the same.